modules/porous_flow/src/materials/PorousFlowDarcyVelocityMaterial.C

```diff
@@ -0,0 +1,114 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PorousFlowDarcyVelocityMaterial.h"
+
+registerMooseObject("PorousFlowApp", PorousFlowDarcyVelocityMaterial);
+
+template <>
+InputParameters
+validParams<PorousFlowDarcyVelocityMaterial>()
+{
+  InputParameters params = validParams<PorousFlowMaterial>();
+  params.addRequiredParam<RealVectorValue>("gravity",
+                                           "Gravitational acceleration vector downwards (m/s^2)");
+  params.addClassDescription("This Material calculates the Darcy velocity for all phases");
+  params.set<bool>("at_nodes") = false;
+  return params;
+}
+
+PorousFlowDarcyVelocityMaterial::PorousFlowDarcyVelocityMaterial(const InputParameters & parameters)
+  : PorousFlowMaterial(parameters),
+    _num_phases(_dictator.numPhases()),
+    _num_var(_dictator.numVariables()),
+    _permeability(getMaterialProperty<RealTensorValue>("PorousFlow_permeability_qp")),
+    _dpermeability_dvar(
+        getMaterialProperty<std::vector<RealTensorValue>>("dPorousFlow_permeability_qp_dvar")),
+    _dpermeability_dgradvar(getMaterialProperty<std::vector<std::vector<RealTensorValue>>>(
+        "dPorousFlow_permeability_qp_dgradvar")),
+    _fluid_density(getMaterialProperty<std::vector<Real>>("PorousFlow_fluid_phase_density_qp")),
+    _dfluid_density_dvar(getMaterialProperty<std::vector<std::vector<Real>>>(
+        "dPorousFlow_fluid_phase_density_qp_dvar")),
+    _fluid_viscosity(getMaterialProperty<std::vector<Real>>("PorousFlow_viscosity_qp")),
+    _dfluid_viscosity_dvar(
+        getMaterialProperty<std::vector<std::vector<Real>>>("dPorousFlow_viscosity_qp_dvar")),
+    _relative_permeability(
+        getMaterialProperty<std::vector<Real>>("PorousFlow_relative_permeability_qp")),
+    _drelative_permeability_dvar(getMaterialProperty<std::vector<std::vector<Real>>>(
+        "dPorousFlow_relative_permeability_qp_dvar")),
+    _grad_p(getMaterialProperty<std::vector<RealGradient>>("PorousFlow_grad_porepressure_qp")),
+    _dgrad_p_dgradvar(getMaterialProperty<std::vector<std::vector<Real>>>(
+        "dPorousFlow_grad_porepressure_qp_dgradvar")),
+    _dgrad_p_dvar(getMaterialProperty<std::vector<std::vector<RealGradient>>>(
+        "dPorousFlow_grad_porepressure_qp_dvar")),
+    _gravity(getParam<RealVectorValue>("gravity")),
+    _darcy_velocity(declareProperty<std::vector<RealVectorValue>>("PorousFlow_darcy_velocity_qp")),
+    _ddarcy_velocity_dvar(declareProperty<std::vector<std::vector<RealVectorValue>>>(
+        "dPorousFlow_darcy_velocity_qp_dvar")),
+    _ddarcy_velocity_dgradvar(
```

Comment:
These data structures are starting to get crazy, have you considered creating a custom type to hold this information that might be a little more straightforward to index and access?

Reply:
Yea, that's a good idea.  Thanks @permcody.  @cpgr and i need to talk about the best design here (part of a different Issue, not this one as we'll need to make pretty substantial changes)